packages/flutter_tools/lib/src/build_system/build_system.dart

```diff
@@ -670,3 +593,131 @@ void verifyOutputDirectories(List<File> outputs, Environment environment, Target
     throw MissingOutputException(missingOutputs, target.name);
   }
 }
+
+/// A node in the build graph.
+class Node {
+  Node(this.target, this.inputs, this.outputs, this.dependencies, Environment environment) {
+    final File stamp = target._findStampFile(environment);
+
+    // If the stamp file doesn't exist, we haven't run this step before and
+    // all inputs were added.
+    if (stamp.existsSync()) {
+      final String content = stamp.readAsStringSync();
+      // Something went wrong writing the stamp file.
+      if (content == null || content.isEmpty) {
+        stamp.deleteSync();
+        // Malformed stamp file, not safe to skip.
+         _dirty = true;
+      } else {
+        final Map<String, Object> values = json.decode(content);
```

Comment:
If the json is non-empty but corrupted somehow, then anything from here down could throw one of several exceptions. Maybe catch and set _dirty to true there too.

Reply:
Updated this to be safe, added test, unnested